rosbag2_transport/test/rosbag2_transport/test_play_timing.cpp

```diff
@@ -190,3 +191,44 @@ TEST_F(PlayerTestFixture, play_ignores_invalid_delay)
   EXPECT_THAT(replay_time, Gt(lower_expected_duration));
   EXPECT_THAT(replay_time, Lt(upper_expected_duration));
 }
+
+TEST_F(PlayerTestFixture, play_respects_starting_time)
+{
+  rclcpp::Duration start_delay(0, static_cast<uint32_t>(5e8));  // 5e8 ns == 0.5 s
+  rclcpp::Duration delay_margin(1, 0);
+
+  // Start 0.5 seconds into the bag
+  play_options_.starting_time = static_cast<rcutils_time_point_value_t>(RCUTILS_S_TO_NS(0.5));
+  // Subtract the delay since we've seeked further into the bag
+  auto lower_expected_duration = message_time_difference - start_delay;
+  auto upper_expected_duration = message_time_difference - start_delay + delay_margin;
+  auto player = std::make_shared<rosbag2_transport::Player>(
+    std::move(reader), storage_options_, play_options_);
+
+  auto start = clock.now();
+  player->play();
+  auto replay_time = clock.now() - start;
+
+  EXPECT_THAT(replay_time, Gt(lower_expected_duration));
+  EXPECT_THAT(replay_time, Lt(upper_expected_duration));
+}
+
+TEST_F(PlayerTestFixture, play_ignores_invalid_starting_time)
```

Comment:
There should be some kind of warning to the user when the starting time is invalid.

Reply:
That makes sense - I guess I overlooked it because the CLI invocation doesn't allow an invalid starting time, but that doesn't stop someone from populating a `rosbag2_transport::PlayOptions` object with the wrong value. Where should this invariant be exercised? It's convenient (not so for the user) to enforce the invariant lazily - i.e. during a `seek`/`play` function - but that may not be the greatest design.

Reply:
I guess if we didn't want to push the negativity check down to the player, then `rosbag2_transport::PlayOptions::starting_time` could be wrapped as a `non_negative<rcutils_time_point_value_t>` with an `operator rcutils_time_point_value_t()` implementation for implicit conversions between the wrapper and the underlying integer type - wrapper in this gist:

[Non-negative signed integral type wrapper](https://gist.github.com/aprotyas/02803a4ade50059285ec8f7badbf2edd)

Reply:
I like the concept that the player options must be correct as constructed, and making the `starting_time` a `non_negative<T>` value allows that - it would cause an error rapidly for the user. However if it's difficult to integrate that cleanly then I'd be happy with checking the options struct's values at the point of use and throwing an error there.

Reply:
In hindsight, it may be hard to integrate this `non_negative<T>`, if for nothing else but for its "sole occurrence" nature in the `PlayOptions` class.
The warning/error for a negative `starting_time` definitely has to go somewhere, but its better to first sort out how the two `starting_time` sources - one from `PlayOptions` and one from `BagMetadata` - live together.

Reply:
I think it makes OK sense to validate `start_offset` in `Player` usage. As it is semantically a Duration, in my view there's nothing wrong theoretically with the data having a negative duration. However, we would consider it a constraint of this specific `Player` implementation that it does not want to start playback before the beginning of the bag (though it would have no technical problem doing so, this use case feels like it's probably not something we would actually want to do in practice).

We're already doing validation of `delay` in `Player::play`, I wouldn't argue against validating `start_offset` there as well.

Reply:
So, as of https://github.com/ros2/rosbag2/pull/931/commits/c79d65cfacb565d6c8b7829ba2bdefce5d61cac0, if the `start_offset` is a negative value, `Player` outputs a message to the warning stream informing the user that it's invalid, but otherwise just ignores the negative value. Thoughts?